Add IO.traverse

leandrob13 opened the pull request:
Attempt to address #11

Waiting for initial thumbs up on the approach to add tests.

ktonga commented on .gitignore:
```diff
@@ -1,3 +1,4 @@
 target
 .sbtopts
 project/.sbt
+.idea
```
Please remove this as it is developer's preference. You should have it in your git globals.

ktonga commented on core/shared/src/main/scala/scalaz/zio/IO.scala:
```diff
@@ -780,6 +782,24 @@ object IO {
+        fn(h).flatMap(a => loopList(t, b += a))
+    }
+
+    loopList(in.toList, cbf(in))
```
You don't need to convert it to list. Also it would be more idiomatic to use `foldLeft` from `TraversableOnce` to do the looping instead of raw recursion.

ktonga replied:
`foldLeft` is implemented in terms of `foreach` which is basically a `while` loop, so it will be easy on the stack :smiley:
If you get stuck you can use the implementation of `Future#traverse` as a reference.

leandrob13 commented:
@ktonga Done! Added a IOSpec since I didn't see a spec for IO. In addition I added specs2-scalacheck. Should we set the `scalafmtOnCompile in ThisBuild := true` option?

jdegoes commented on core/jvm/src/test/scala/scalaz/zio/IOSpec.scala:
```diff
@@ -0,0 +1,45 @@
+
+import scala.util.Try
+
+class IOSpec extends Specification with GenIO with RTS with ScalaCheck {
```
It's a nice idea in theory, I am not sure how you can decide what is primitive and not???

leandrob13 replied:
Sorry, I am just getting familiar with the project. Should I move all the tests to RTSSpec then?

ktonga replied:
I was thinking things like testing trampolining, error handling, finalization, etc belong in `RTSSpec` and things like testing all the several combinators accepting traversables recently added belong in `IOSpec`. The rest we can see.

ktonga replied:
@leandrob13 My vote is to keep `IOSpec`, but let's wait for @jdegoes to decide.

leandrob13 replied:
Seems ok to me, although I didn't catch it at first (to be honest I don't know what RTS stands for) and that is why I created IOSpec.

ktonga replied:
I don't blame you, that's the right instinct. That's why I want to keep it.

leandrob13 replied:
Run Time System, I did miss it in the javadoc.
```
 * This trait provides a high-performance implementation of a runtime system for
 * the `IO` monad on the JVM.
```

ktonga commented:
> Added a IOSpec since I didn't see a spec for IO. In addition I added specs2-scalacheck

I like having `IOSpec` but we were putting all `IO`-related cases in `RTSSpect`. +100 re: scalacheck

> Should we set the scalafmtOnCompile in ThisBuild := true option?

Still to be decided for all the scalaz projects. Not as part of this PR for sure.

leandrob13 commented on core/shared/src/main/scala/scalaz/zio/IO.scala:
```diff
@@ -780,6 +782,16 @@ object IO {
+  def traverse[E, A, B, M[X] <: TraversableOnce[X]](
+    in: M[A]
+  )(fn: A => IO[E, B])(implicit cbf: CanBuildFrom[M[A], B, M[B]]): IO[E, M[B]] =
+    in.foldLeft(point[E, mutable.Builder[B, M[B]]](cbf(in)))((io, b) => io.zipWith(fn(b))(_ += _))
```
Nope, I ger:

```
type mismatch;
[error]  found   : scalaz.zio.IO[E,B]
[error]  required: scalaz.zio.IO[Nothing,?]
```

regiskuckaertz commented:
Note you could use `.par` for potentially great savings but at the cost of losing the ordering guarantee (if the `M[_]` has that property).

regiskuckaertz commented:
Since `sequence = traverse(identity)`, it may not be worth a separate PR?

ktonga commented:
IMO traverse has to sequence the actions and respect the iteration order of the traversable, we can have some other method for a non-deterministic approach. So current implementation is ok.
+1 to adding `sequence` as part of this PR

jdegoes commented:
This should probably be called `traverseSeq` to emphasize the order is preserved.
